File: src/main/scala/MaximumFrequencyStack.java
import java.util.*;

public class MaximumFrequencyStack {
    private Map<Integer,Integer> freq = new HashMap<>();
    private Map<Integer, Stack<Integer>> group = new HashMap<>();
    private int maxFreq;

    public void push(int x) {
        int f = freq.getOrDefault(x, 0);
        f += 1;
        freq.put(x, f);
        if(f > maxFreq)
            maxFreq = f;
        group.computeIfAbsent(f, z -> new Stack()).push(x);
    }


    public int pop() {
        Stack<Integer> stack = group.get(maxFreq);
        int max = stack.pop();
        freq.put(max, freq.get(max) - 1);
        if(stack.isEmpty()){
            group.remove(maxFreq);
            maxFreq -= 1;
        }
        return max;
    }


    public static void main(String ... args){
        MaximumFrequencyStack stack = new MaximumFrequencyStack();
        //[5,7,5,7,4,5]
        stack.push(5);
        stack.push(7);
        stack.push(5);
        stack.push(7);
        stack.push(4);
        stack.push(5);
        for(int i = 0; i< 6;i++){
            System.out.println(stack.pop());
        }
    }

}

File: src/main/scala/MaximumFrequencyStack.py
from collections import defaultdict

class FreqStack:

    def __init__(self):
        self.max_freq = 0
        self.item_to_freq = defaultdict(int)
        self.freq_to_item = defaultdict(list)

    def push(self, val: int) -> None:
        self.item_to_freq[val] += 1
        freq = self.item_to_freq[val]
        self.max_freq = max(freq, self.max_freq)
        self.freq_to_item[freq].append(val)

    def pop(self) -> int:
        val = self.freq_to_item[self.max_freq].pop()
        self.item_to_freq[val] = self.max_freq - 1
        if len(self.freq_to_item[self.max_freq]) == 0:
            del self.freq_to_item[self.max_freq]
            self.max_freq -= 1
        return val
#"push", "push", "push", "push", "push", "push", "pop", "pop", "pop", "pop"]
#[[], [5], [7], [5], [7], [4], [5], [], [], [], []]
freqStack = FreqStack()
freqStack.push(5)
freqStack.push(7)
freqStack.push(5)
freqStack.push(7)
freqStack.push(4)
freqStack.push(5)
print(freqStack.pop())#5
print(freqStack.pop())#7
print(freqStack.pop())#5
print(freqStack.pop())#4